src/main/java/com/couchbase/lite/Database.java

```diff
@@ -1152,27 +1149,54 @@ private void markBlobstoreMigrated() {
      */
     @InterfaceAudience.Private
     public boolean close() {
-        if(!open) {
+        if (!open) {
             return false;
         }
 
-        if(views != null) {
+        if (views != null) {
             for (View view : views.values()) {
                 view.databaseClosing();
             }
         }
         views = null;
 
-        if(activeReplicators != null) {
-            for(Replication replicator : activeReplicators) {
+        if (activeReplicators != null) {
+            List<CountDownLatch> latches = new ArrayList<CountDownLatch>();
+            for (Replication replicator : activeReplicators) {
+                // handler to check if the replicator stopped
+                final CountDownLatch latch = new CountDownLatch(1);
+                replicator.addChangeListener(new Replication.ChangeListener() {
+                    @Override
+                    public void changed(Replication.ChangeEvent event) {
+                        if (event.getSource().getStatus() == Replication.ReplicationStatus.REPLICATION_STOPPED) {
+                            latch.countDown();
+                        }
+                    }
+                });
+                latches.add(latch);
+
+                // ask replicator to stop
                 replicator.databaseClosing();
             }
+
```

Comment:
Response for question 1:
- In general, Database class is not thread safe. Only `open()` method is `synchronized` method to protected SQLite database. From the code, I consider that `close()` method is not necessary to be synchronized method. For safety, we can apply `synchronized`.

- Regarding SQLiteDatabase (SQLite), it provides a locking mechanism by default. So storage is thread safe.
http://stackoverflow.com/questions/6675240/is-sqlite-database-instance-thread-safe
http://developer.android.com/reference/android/database/sqlite/SQLiteDatabase.html#setLockingEnabled%28boolean%29